src/phoenix/auth.py

```diff
@@ -33,16 +61,165 @@ def validate_password_format(password: str) -> None:
+    """
 
+    length: int
+    special_chars: bool = False
+    digits: bool = False
+    upper_case: bool = False
+    lower_case: bool = False
 
+    def validate(self, password: str) -> None:
+        """
+        Validates the password against the requirements.
+
+        Args:
+            password (str): the password to validate
+        Returns:
+            None
+        Raises:
+            ValueError: if the password does not meet the requirements
+        """
+        if not password:
+            raise ValueError("Password must be non-empty")
+        if any(char.isspace() for char in password):
+            raise ValueError("Password must not contain whitespace characters")
+        if not password.isascii():
+            raise ValueError("Password must contain only ASCII characters")
+        err_msg = []
+        if len(password) < self.length:
+            err_msg.append(f"must be at least {self.length} characters long")
+        if self.special_chars and not any(char in "!@#$%^&*()_+" for char in password):
+            err_msg.append("at least one special character")
+        if self.digits and not any(char.isdigit() for char in password):
+            err_msg.append("at least one digit")
+        if self.upper_case and not any(char.isupper() for char in password):
+            err_msg.append("at least one uppercase letter")
+        if self.lower_case and not any(char.islower() for char in password):
+            err_msg.append("at least one lowercase letter")
+        if not err_msg:
+            return
+        try:
+            from faker import Faker  # type: ignore[import-not-found,unused-ignore]
+
+            suggested_password = Faker().unique.password(**asdict(self))
+        except BaseException:
+            suggested_password = ""
+        if len(err_msg) > 1:
+            err_text = "Password " + ", ".join(err_msg[:-1]) + ", and " + err_msg[-1]
+        else:
+            err_text = f"Password {err_msg[0]}"
+        if suggested_password:
+            err_text = f"{err_text} (e.g. `{suggested_password}`)"
+        raise ValueError(err_text)
+
+
+"""The name of the header for token-based authentication"""
 EMAIL_PATTERN = re.compile(r"^[^@\s]+@[^@\s]+[.][^@\s]+\Z")
+"""The regular expression pattern for a valid email address."""
 NUM_ITERATIONS = 10_000
+"""The number of iterations to use for the PBKDF2 key derivation function."""
 MIN_PASSWORD_LENGTH = 4
+"""The minimum length of a password."""
+PASSWORD_REQUIREMENTS = _PasswordRequirements(MIN_PASSWORD_LENGTH)
+"""The requirements for a valid password."""
+REQUIREMENTS_FOR_PHOENIX_SECRET = _PasswordRequirements(32, True, True, True, True)
+"""The requirements for the Phoenix secret key."""
+JWT_ALGORITHM = "HS256"
+"""The algorithm to use for the JSON Web Token."""
+PHOENIX_ACCESS_TOKEN_COOKIE_NAME = "phoenix-access-token"
+PHOENIX_ACCESS_TOKEN_COOKIE_MAX_AGE = timedelta(days=31)
+
+
+class Issuer(Enum):
+    API_KEY = auto()
+    SESSION = auto()
```

Comment:
this doesn't really match my understanding of issuer: https://www.ietf.org/archive/id/draft-meyerzuselhausen-oauth-iss-auth-resp-00.html#section-2.1

Let's avoid overloading this term. 

Let's also avoid the word session as it has a lot of connotation about propagating state from client / server. We should just call it what it is - access_token authentication.

Reply:
I think that document is OAuth specific, but the RFC permits application of this concept that is application-specific.